### PR TITLE
Switch to Barclay command line parser.

### PR DESCRIPTION
**Pull request by cmnbroad**
https://github.com/broadinstitute/gatk/issues/2210. Removed ArgumentCollectionDefinition, and migrated the rest of the CLP annotations to Barclay. Mostly package name changes and deletions. CommandLineException is no longer derived from UserException.

### CODE REVIEW

**Comment by droazen** [comment]
For @lbergelson

**Comment by vdauwera** [comment]
Exciting!

**Comment by lbergelson** [comment]
👍 Looks good to me.  Did you want to try to switch to the release version, or should we merge this as is?  

I didn't know even know we had redundant `hidden` / `hiddenOption` tags...  both unused.

**Comment by magicDGS** [comment]
@cmnbroad, I've already change my toolkits to use the append collection feature (https://github.com/broadinstitute/barclay/pull/3). Is it going to be included in this PR?

**Comment by cmnbroad** [comment]
@magicDGS Yes; I won't merge this until we move up to a released Barclay with that change in it.

**Comment by magicDGS** [comment]
Thanks for the answer, @cmnbroad!

File: src/main/java/org/broadinstitute/hellbender/cmdline/SpecialArgumentsCollection.java [hunk 1]
```diff
@@ -4,12 +4,13 @@
...
 import java.util.List;
 
+import org.broadinstitute.barclay.argparser.Argument;
 
 /**
  * This collection is for arguments that require special treatment by the arguments parser itself.
  * It should not grow beyond a very short list.
  */
-public final class SpecialArgumentsCollection implements ArgumentCollectionDefinition {
+public final class SpecialArgumentsCollection {
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/cmdline/SpecialArgumentsCollection.java]: Shouldn't this implement `Serializable`?

File: src/main/java/org/broadinstitute/hellbender/Main.java [hunk 2]
```diff
@@ -40,7 +41,7 @@
     private static final String BOLDRED = "\u001B[1m\u001B[31m";
 
     /**
-     * exit value when an issue with the commandline is detected, ie {@link UserException.CommandLineException}.
+     * exit value when an issue with the commandline is detected, ie CommandLineException.
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/Main.java]: Couldn't this keep the `{@link CommandLineException}` javadoc annotation?

File: src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java [hunk 3]
```diff
@@ -22,7 +23,7 @@
 /**
  * A GATKCommandLinePluginDescriptor for ReadFilter plugins
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java]: GATKCommandLinePluginDescriptor -> CommandLinePluginDescriptor

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/IntervalArgumentCollection.java [hunk 4]
```diff
@@ -4,21 +4,23 @@
...
 import org.broadinstitute.hellbender.exceptions.UserException;
 import org.broadinstitute.hellbender.utils.*;
 
+import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.List;
 
 /**
- * Intended to be used as an {@link org.broadinstitute.hellbender.cmdline.ArgumentCollection} for specifying intervals at the command line.
+ * Intended to be used as an @ArgumentCollection for specifying intervals at the command line.
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/IntervalArgumentCollection.java]: Couldn't this keep the `{@link ArgumentCollection}` javadoc annotation?

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/IntervalArgumentCollection.java [hunk 5]
```diff
@@ -126,13 +128,13 @@ private void parseIntervals(final GenomeLocParser genomeLocParser) {
...
             } catch( UserException.EmptyIntersection e) {
-                throw new UserException.BadArgumentValue("-L, --interval_set_rule", getIntervalStrings()+","+intervalSetRule, "The specified intervals had an empty intersection");
+                throw new CommandLineException.BadArgumentValue("-L, --interval_set_rule", getIntervalStrings()+","+intervalSetRule, "The specified intervals had an empty intersection");
             }
         }
 
         final GenomeLocSortedSet excludeSortedSet = IntervalUtils.loadIntervals(excludeIntervalStrings, IntervalSetRule.UNION, intervalMerging, intervalExclusionPadding, genomeLocParser);
         if ( excludeSortedSet.contains(GenomeLoc.UNMAPPED) ) {
-            throw new UserException("-XL unmapped is not currently supported");
+            throw new CommandLineException("-XL unmapped is not currently supported");
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/IntervalArgumentCollection.java]: This is changing the behaviour for exception handling, no? Before it wasn't a `CommandLineException`.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OpticalDuplicatesArgumentCollection.java [hunk 6]
```diff
@@ -1,19 +1,16 @@
...
-import java.util.List;
+import java.io.Serializable;
 
 
 /**
  * An argument collection for use with tools that mark optical
  * duplicates.
  */
-public final class OpticalDuplicatesArgumentCollection implements ArgumentCollectionDefinition {
+public final class OpticalDuplicatesArgumentCollection {
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OpticalDuplicatesArgumentCollection.java]: Shouldn't this implement `Serializable`?

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalVariantInputArgumentCollection.java [hunk 7]
```diff
@@ -2,21 +2,21 @@
...
 /**
  * An argument collection for use with tools that accept zero or more input files containing VariantContext records
  * (eg., VCF files).
  */
-public final class OptionalVariantInputArgumentCollection implements ArgumentCollectionDefinition {
+public final class OptionalVariantInputArgumentCollection implements Serializable {
     private static final long serialVersionUID = 1L;
 
-    @Argument(fullName = StandardArgumentDefinitions.VARIANT_LONG_NAME, shortName = StandardArgumentDefinitions.VARIANT_SHORT_NAME, doc = "Variants file", optional = true)
+    @org.broadinstitute.barclay.argparser.Argument(fullName = StandardArgumentDefinitions.VARIANT_LONG_NAME, shortName = StandardArgumentDefinitions.VARIANT_SHORT_NAME, doc = "Variants file", optional = true)
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalVariantInputArgumentCollection.java]: I guess that this diff could be removed without specifying the package name.

File: src/main/java/org/broadinstitute/hellbender/cmdline/SpecialArgumentsCollection.java (hunk 1 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/cmdline/SpecialArgumentsCollection.java]: Yes.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/IntervalArgumentCollection.java (hunk 5 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/IntervalArgumentCollection.java]: Yes, this one didn't need to change.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OpticalDuplicatesArgumentCollection.java (hunk 6 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OpticalDuplicatesArgumentCollection.java]: Yes.

**Pull request merged**